onnxruntime/core/providers/cpu/tensor/copy.cc

```diff
@@ -0,0 +1,151 @@
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/copy.h"
+#include "core/platform/threadpool.h"
+
+namespace onnxruntime {
+
+std::vector<int64_t> StridesForTensor(const Tensor& dst) {
+  std::vector<int64_t> strides(dst.Shape().NumDimensions());
+  for (std::size_t i = 0; i < dst.Shape().NumDimensions(); i++) {
+    strides[i] = 1;
+    for (std::size_t j = i + 1; j < dst.Shape().NumDimensions(); j++) {
+      strides[i] *= dst.Shape()[j];
+    }
+  }
+  return strides;
+}
+
+Status DispatchStridedCopy(concurrency::ThreadPool* thread_pool,
+                           Tensor& dst,
+                           std::ptrdiff_t dst_offset,
+                           const std::vector<int64_t> dst_strides,
+                           const TensorShape& copy_shape,
+                           const Tensor& src,
+                           const std::vector<int64_t> src_strides) {
+  ORT_RETURN_IF_NOT(dst.DataType() == src.DataType(), "src and dst types must match");
+  ORT_RETURN_IF_NOT(dst_strides.size() == src_strides.size() && src_strides.size() == copy_shape.NumDimensions(), "src and dst must have same shape");
```

Comment:
If these tests fail, is it a bug in ORT, or an error in the graph the user provided?  If it's an ORT internal bug then I'd use assert(...) to add the checking rather than exposing an error to the user.